From 325323e439d476899c395144b154e86e6ca87aeb Mon Sep 17 00:00:00 2001
From: LeiZhou-97 <lei.zhou@intel.com>
Date: Tue, 29 Aug 2023 16:52:06 +0800
Subject: [PATCH 11/18] Enable Tuning of the guest I/O APIC mode.

Set the qemu driver attribute for TDX VM by default.

Signed-off-by: LeiZhou-97 <lei.zhou@intel.com>
---
 .../virtwrap/api/deepcopy_generated.go        | 21 +++++++++++++++++++
 pkg/virt-launcher/virtwrap/api/schema.go      |  5 +++++
 .../virtwrap/converter/converter.go           |  3 +++
 3 files changed, 29 insertions(+)

diff --git a/pkg/virt-launcher/virtwrap/api/deepcopy_generated.go b/pkg/virt-launcher/virtwrap/api/deepcopy_generated.go
index b5cb529e2f36..7635f35bd71b 100644
--- a/pkg/virt-launcher/virtwrap/api/deepcopy_generated.go
+++ b/pkg/virt-launcher/virtwrap/api/deepcopy_generated.go
@@ -1351,6 +1351,22 @@ func (in *FeatureHyperv) DeepCopy() *FeatureHyperv {
 	return out
 }
 
+// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
+func (in *FeatureIOAPIC) DeepCopyInto(out *FeatureIOAPIC) {
+	*out = *in
+	return
+}
+
+// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new FeatureIOAPIC.
+func (in *FeatureIOAPIC) DeepCopy() *FeatureIOAPIC {
+	if in == nil {
+		return nil
+	}
+	out := new(FeatureIOAPIC)
+	in.DeepCopyInto(out)
+	return out
+}
+
 // DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
 func (in *FeatureKVM) DeepCopyInto(out *FeatureKVM) {
 	*out = *in
@@ -1484,6 +1500,11 @@ func (in *Features) DeepCopyInto(out *Features) {
 		*out = new(FeatureState)
 		**out = **in
 	}
+	if in.IOAPIC != nil {
+		in, out := &in.IOAPIC, &out.IOAPIC
+		*out = new(FeatureIOAPIC)
+		**out = **in
+	}
 	return
 }
 
diff --git a/pkg/virt-launcher/virtwrap/api/schema.go b/pkg/virt-launcher/virtwrap/api/schema.go
index 465c6c6c10fe..19eddb33130a 100644
--- a/pkg/virt-launcher/virtwrap/api/schema.go
+++ b/pkg/virt-launcher/virtwrap/api/schema.go
@@ -298,6 +298,7 @@ type Features struct {
 	KVM        *FeatureKVM        `xml:"kvm,omitempty"`
 	PVSpinlock *FeaturePVSpinlock `xml:"pvspinlock,omitempty"`
 	PMU        *FeatureState      `xml:"pmu,omitempty"`
+	IOAPIC     *FeatureIOAPIC     `xml:"ioapic,omitempty"`
 }
 
 type FeatureHyperv struct {
@@ -322,6 +323,10 @@ type FeatureSpinlocks struct {
 	Retries *uint32 `xml:"retries,attr,omitempty"`
 }
 
+type FeatureIOAPIC struct {
+	Driver string `xml:"driver,attr,omitempty"`
+}
+
 type SyNICTimer struct {
 	Direct *FeatureState `xml:"direct,omitempty"`
 	State  string        `xml:"state,attr,omitempty"`
diff --git a/pkg/virt-launcher/virtwrap/converter/converter.go b/pkg/virt-launcher/virtwrap/converter/converter.go
index 26afcbe15732..b9aa491f1048 100644
--- a/pkg/virt-launcher/virtwrap/converter/converter.go
+++ b/pkg/virt-launcher/virtwrap/converter/converter.go
@@ -1097,6 +1097,9 @@ func Convert_v1_Features_To_api_Features(source *v1.Features, features *api.Feat
 			State: boolToOnOff(source.Pvspinlock.Enabled, true),
 		}
 	}
+	if c.UseLaunchSecurityTDX {
+		features.IOAPIC = &api.FeatureIOAPIC{Driver: "qemu"}
+	}
 	return nil
 }
 
-- 
2.45.0

